File: scml_mat/pps.m
function id = pps(knn, rnn, order)
%   Plum Pudding Sampling (PPS)
%   This function returns the ID of landmarks. 
%
%   Parameters are:
%
%   'knn'   - A N by k matrix. Each row represents the KNN of each point. 
%   'rnn'   - A vector of length N. Each row represents the RNN of each point. 
%   'order' - A positive integer specifying the order of KNN. Once a landmark is selected, its 
%             KNN will be removed from the queue. PPS supports the removal of multi-order KNN. 

id = [];
[~, id_sort] = sort(rnn,'descend');
while ~isempty(id_sort)
    id = [id;id_sort(1)];
    rm_pts = [id_sort(1)];
    for i=1:order
        temp_pts = knn(rm_pts,:);
        rm_pts = [rm_pts;temp_pts(:)];
    end
    rm_pts = unique(rm_pts);
    [~,rm_id] = ismember(rm_pts,id_sort);
    id_sort(nonzeros(rm_id)) = [];    
end

File: sude_py/sude/pps.py
def pps(knn, rnn, order):
    """
    Plum Pudding Sampling (PPS)
    This function returns the ID of landmarks.

    Parameters are:

    'knn'      - A N by k matrix. Each row represents the KNN of each point.
    'rnn'      - A vector of length N. Each row represents the RNN of each point.
    'order'    - A positive integer specifying the order of KNN. Once a landmark is selected, its KNN will be removed
                 from the queue. PPS supports the removal of multi-order KNN.

    """
    id_samp = []
    id_sort = sorted(range(len(rnn)), key=lambda k: rnn[k], reverse=True)
    while len(id_sort) > 0:
        id_samp.append(id_sort[0])
        rm_pts = [id_sort[0]]
        for _ in range(order):
            rm_pts.extend(knn[rm_pts].flatten().tolist())
        rm_pts = set(rm_pts)
        id_sort = [point_id for point_id in id_sort if point_id not in rm_pts]
    return id_samp
